### DISCUSSION TITLE
Trouble getting a simple test-case to fail (or succeed, so you want)

### DISCUSSION THREAD

**Discussion by NicolasT** [IntersectMBO/io-sim#61] [Q&A] [answered]
I found out about `io-sim` yesterday, and I'm quite interested to use it for some project. I believe this tool could be a game-changer in the Haskell testing landscape.

Unless I missed something, documentation is somewhat sparse, so I tried to get a small example up-and-running by looking at Haddocks and source-code. However, I'm unable to get it to work.

Or maybe I have an incorrect understanding of what the tool is supposed to do...

The testcase I wrote is quite simple: it create a `TVar` initialized with `0`, then spawns 4 threads: 3 of them will `atomically` put `2`, `3` or `4` in said `TVar` once they read `1` from it (otherwise `retry`), one will (non-atomically) write `1` into said `TVar`, then (`atomically`) read from it until it finds a non-`1` value. This is, then, the return value of the testcase.

I wrote two properties, one in plain `IO`, one using `IOSim`. They both run the testcase (either in `IO`, or as a trace using `exploreSimTrace`/`traceResult`), then assert the result value equals `4`.

If I'm not mistaken, the return value *can* be `4`, but it could as well be `2` or `3`. Indeed, the `IO` property fails after some iterations (it may succeed, of course, in the end, we're testing a race-condition...). However, the `IOSim`-based test always succeeds, whilst I was expecting it to (always, even, given the relatively small state space?) fail.

Any pointers?

`Main.hs`:
```haskell
{-# LANGUAGE TypeApplications #-}

module Main where

import Control.Applicative ((<|>))

import Control.Monad.Class.MonadAsync (MonadAsync, Concurrently(..), runConcurrently)
import Control.Monad.Class.MonadSTM.Internal (MonadSTM, atomically, check, newTVarIO, readTVar, writeTVar)
import Control.Monad.Class.MonadTest (MonadTest, exploreRaces)
import Control.Monad.Class.MonadTimer (MonadTimer)

import Test.Hspec (describe, hspec)
import Test.Hspec.QuickCheck (prop)
import Test.QuickCheck (Property, (===), counterexample)
import Test.QuickCheck.Monadic (monadicIO, run, stop)

import Control.Monad.IOSim (exploreSimTrace, traceResult)

main :: IO ()
main = hspec $ do
    describe "Testcase" $ do
        prop "in IO" propIO
        prop "in IOSim" propIOSim

-- Run the testcase using 'real' STM in IO, and check whether the result
-- equals 4 (it could, but definitely shouldn't always).
propIO :: Property
propIO = monadicIO @() $ do
    r <- run testCase
    -- This is supposed to fail quite regularly
    stop (r === 4)

-- Run the testcase using IOSim, and check whether the result equals 4.
-- It sometimes could, but shouldn't for every possible interleaving of
-- thread actions (as witnessd by the IO-based test).
propIOSim :: Property
propIOSim = exploreSimTrace id testCase $ \_ trace ->
    case traceResult False trace of
        Right result -> result === 4
        Left e -> counterexample (show e) False

-- This test case is quite simple: it's based around a TVar whose initial
-- value is 0. Then, 4 threads are spawned: three of them will, atomically,
-- write some value (2, 3 or 4) in the TVar once they read the value 1 from
-- it. These three threads race: once one of them was able to write its value,
-- others will no longer read 1 from the TVar.
--
-- A fourth thread will write 1 into the TVar, and then (in another transaction)
-- wait for the value found in the TVar to be non-1 (at this point, this should
-- be one of 2, 3 or 4). It then returns this value.
testCase :: (MonadAsync m, MonadSTM m, MonadTest m, MonadTimer m) => m Int
testCase = do
    v <- newTVarIO (0 :: Int)

    runConcurrently $ (,) <$> (  Concurrently (exploreRaces >> writer v 2) 
                             <|> Concurrently (exploreRaces >> writer v 3)
                             <|> Concurrently (exploreRaces >> writer v 4)
                              )
                           *> Concurrently (exploreRaces >> driver v)
  where
    writer v i = atomically $ do
        -- Wait for driver to change value
        v' <- readTVar v
        check (v' == 1)
        -- Write "our" value
        writeTVar v i

    driver v = do
        -- Let (one of) the other threads continue
        atomically $ writeTVar v 1

        atomically $ do
            -- Wait until some other thread wrote some other value
            v' <- readTVar v
            check (v' /= 1)
            return v'
```

`cabal.project`:
```
packages: .

source-repository-package
  type: git
  location: https://github.com/input-output-hk/io-sim.git
  tag: 9b06c704a4656aaa6b5c1f080658426a5e469f4f
  subdir: io-classes io-sim strict-stm
```

`sim.cabal`:
```
Cabal-Version: 3.4
Build-Type: Simple

Name: sim
Version: 0.0.1.0
Synopsis: An experiment with io-sim
Description: A testcase using io-sim to play with its functionality
Author: Nicolas Trangez
Maintainer: ikke@nicolast.be
Copyright: (c) 2022, Nicolas Trangez
License: GPL-3.0-or-later

Category: Testing

Test-Suite sim-test
  Main-Is: Main.hs
  Type: exitcode-stdio-1.0
  Build-Depends:        base ^>=4.17.0.0
                      , hspec ^>=2.10.7
                      , io-classes ^>=0.4.0.0
                      , io-sim ^>=0.4.0.0
                      , QuickCheck ^>=2.14
```

**Comment by coot** [comment, marked as answer]
By default all the threads are not _racy_, you need to execute `exploreRaces` in the main thread (all threads forked from a _racy_ thread will take part in the schedule exploration).  I think this is explained [here](https://github.com/input-output-hk/io-sim/blob/main/io-sim/how-to-use-IOSimPOR.md).

The default schedule of `IOSimPOR` is a round-robin schedule, I guess the property is just true in that schedule.

**Reply by NicolasT** [reply]
Thanks for geting back! I had read the how-to, but misinterpreted it: I thought (as done in the original code above) I had to call `exploreRaces` in every thread part of the "system", and then indeed threads forked further down the line would also be system threads. However, after re-reading the how-to, indeed I need to call it from the "main" thread directly.

Given this, also changing the property from checking equality to 4 to equality to 2, results in both the `IO` as well as the `IOSim` version to fail:

```
Testcase
  in IO [✘]
  in IOSim [✘]

Failures:

  Main.hs:21:5: 
  1) Testcase in IO
       Falsified (after 2 tests):
         4 /= 2

  To rerun use: --match "/Testcase/in IO/"

  Main.hs:22:5: 
  2) Testcase in IOSim
       Falsified (after 1 test):
         Schedule control: ControlAwait [ScheduleMod (RacyThreadId [1,1,1,1],1) ControlDefault [(RacyThreadId [1,1,1],3),(RacyThreadId [1,1,1],4),(RacyThreadId [1,1,1],5),(RacyThreadId [1,1,1],6),(RacyThreadId [1,1,1,2],0),(RacyThreadId [1,1,1],7),(RacyThreadId [1,1,1],8),(RacyThreadId [1,1],3),(RacyThreadId [1,1],4),(RacyThreadId [1,1],5),(RacyThreadId [1,1],6),(RacyThreadId [1,1,2],0),(RacyThreadId [1,1],7),(RacyThreadId [1,1],8),(RacyThreadId [1,1,2],1)]]
         RacyThreadId [1,2] (driver) delayed at time Time 0s
           until after:
             RacyThreadId [1,1]
             RacyThreadId [1,1,1]
             RacyThreadId [1,1,1,1]
             RacyThreadId [1,1,1,2]
             RacyThreadId [1,1,2]
             RacyThreadId [1,1,2] (writer4)
         
         4 /= 2

  To rerun use: --match "/Testcase/in IOSim/"

Randomized with seed 993392668

Finished in 0.0297 seconds
2 examples, 2 failures
```

Not sure how to interpret the trace and other output, for now. Also it seems not all thread labels are displayed? Will look into that.

For reference, changed `testCase` code:
```haskell
testCase :: (MonadAsync m, MonadLabelledSTM m, MonadTest m, MonadTimer m) => m Int
testCase = do
  flip withAsync wait $ do
    exploreRaces
    labelThisThread "main"

    v <- newTVarIO (0 :: Int)
    labelTVarIO v "cell"

    runConcurrently $
      (,)
        <$> ( Concurrently (labelThisThread "writer2" >> writer v 2)
                <|> Concurrently (labelThisThread "writer3" >> writer v 3)
                <|> Concurrently (labelThisThread "writer4" >> writer v 4)
            )
          *> Concurrently (labelThisThread "driver" >> driver v)
```

**Reply by coot** [reply]
My hypothesis why other threads are not labelled is because `lableThisThread` in them was delayed. If you use `Concurrently (labelThisThread "writerX" >> threadDelay 1 >> writer v x)` then they all should be labelled and the race will happen in the same slot time, so should be discovered by `IOSimPOR`.

I only have an educated guess why my fix worked: probably the `TVar`s on which the main thread is blocked are needed to discover the race between the other threads, which might be because when one of the `writer`'s succeeds that thread `Async` (which is backed by a `TVar` will be filled), the main thread will still be blocked on the other `Async`s.  This makes events in the main thread needed to discover race between the other threads (that's the hand-wavy part :wink:).
